ompi/mpi/c/isendrecv_replace.c

```diff
@@ -0,0 +1,239 @@
+{
+    ompi_isendrecv_replace_context_t *context =
+        (ompi_isendrecv_replace_context_t *) request->context;
+
+    /*
+     * Copy the status from the receive side of the sendrecv request?
+     * But what if the send failed?
+     *
+     * Probably need to bring up in the MPI forum.  
+     */
+    
+    OMPI_COPY_STATUS(&request->super.req_status, 
+                     context->statuses[0], false);
+
+    return OMPI_SUCCESS;
+}
+
+
+int MPI_Isendrecv_replace(void * buf, int count, MPI_Datatype datatype,
+                         int dest, int sendtag, int source, int recvtag,
+                         MPI_Comm comm, MPI_Request *request)
+
+{
+    int rc = MPI_SUCCESS;
+    size_t max_data;
+    uint32_t iov_count;
+    ompi_comm_request_t *crequest = NULL;
+    ompi_isendrecv_replace_context_t *context = NULL;
+    ompi_request_t *subreq[2];
+    int nreqs = 0;
+
+    SPC_RECORD(OMPI_SPC_SENDRECV_REPLACE, 1);
+
+    MEMCHECKER(
+               memchecker_datatype(datatype);
+               memchecker_call(&opal_memchecker_base_isdefined, buf, count, datatype);
+               memchecker_comm(comm);
+               );
+
+    if ( MPI_PARAM_CHECK ) {
+        rc = MPI_SUCCESS;
+        OMPI_ERR_INIT_FINALIZE(FUNC_NAME);
+        OMPI_CHECK_DATATYPE_FOR_RECV(rc, datatype, count);
+
+        if (ompi_comm_invalid(comm)) {
+            return OMPI_ERRHANDLER_NOHANDLE_INVOKE(MPI_ERR_COMM, FUNC_NAME);
+        } else if (dest != MPI_PROC_NULL && ompi_comm_peer_invalid(comm, dest)) {
+            rc = MPI_ERR_RANK;
+        } else if (sendtag < 0 || sendtag > mca_pml.pml_max_tag) {
+            rc = MPI_ERR_TAG;
+        } else if (source != MPI_PROC_NULL && source != MPI_ANY_SOURCE && ompi_comm_peer_invalid(comm, source)) {
+            rc = MPI_ERR_RANK;
+        } else if (((recvtag < 0) && (recvtag !=  MPI_ANY_TAG)) || (recvtag > mca_pml.pml_max_tag)) {
+            rc = MPI_ERR_TAG;
+        } else if (request == NULL) {
+            rc = MPI_ERR_REQUEST;
+        }
+        OMPI_ERRHANDLER_CHECK(rc, comm, rc, FUNC_NAME);
+    }
+
+    /* simple case */
+    if ( source == MPI_PROC_NULL || dest == MPI_PROC_NULL || count == 0 ) {
+        rc = PMPI_Isendrecv(buf, count, datatype, dest, sendtag, buf, count, datatype, source, recvtag, comm, request);
+        return rc;
+    }
+
+    /**
+     * If we look for an optimal solution, then we should receive the data into a temporary buffer
+     * and once the send completes we would unpack back into the original buffer. However, if the
+     * sender is unknown, this approach can only be implementing by receiving with the recv datatype
+     * (potentially non-contiguous) and thus the allocated memory will be larger than the size of the
+     * datatype. A simpler, but potentially less efficient approach is to work on the data we have
+     * control of, aka the sent data, and pack it into a contiguous buffer before posting the receive.
+     * Once the send completes, we free it.
+     */
```

Comment:
Not really correct. If the build is homogeneous then you can just use the local size. For a heterogeneous build check the mca_coll_base_alltoallv_intra_basic_inplace function to see how this can be handled efficiently.

Reply:
this is residue from starting from sendrecv_replace.c   just deleted the whole comment in this file.